mlflow/langchain/databricks_dependencies.py

```diff
@@ -0,0 +1,185 @@
+    else:
+        d[key] = value
+    return d
+
+
+def _extract_databricks_dependencies_from_retriever(retriever, dependency_dict) -> Dict[str, str]:
+    import langchain
+
+    if hasattr(retriever, "vectorstore") and hasattr(retriever.vectorstore, "embeddings"):
+        vectorstore = retriever.vectorstore
+        embeddings = vectorstore.embeddings
+
+        if isinstance(vectorstore, langchain.vectorstores.DatabricksVectorSearch):
+            index = vectorstore.index
+            _assign_value_or_append_to_list(
+                dependency_dict, _DATABRICKS_VECTOR_SEARCH_INDEX_NAME_KEY, index.name
+            )
+            _assign_value_or_append_to_list(
+                dependency_dict, _DATABRICKS_VECTOR_SEARCH_ENDPOINT_NAME_KEY, index.endpoint_name
+            )
+
+        if isinstance(embeddings, langchain.embeddings.DatabricksEmbeddings):
+            _assign_value_or_append_to_list(
+                dependency_dict, _DATABRICKS_EMBEDDINGS_ENDPOINT_NAME_KEY, embeddings.endpoint
+            )
+        elif (
+            hasattr(vectorstore, "_is_databricks_managed_embeddings")
+            and callable(getattr(vectorstore, "_is_databricks_managed_embeddings"))
+            and vectorstore._is_databricks_managed_embeddings()
+        ):
+            _assign_value_or_append_to_list(
+                dependency_dict,
+                _DATABRICKS_EMBEDDINGS_ENDPOINT_NAME_KEY,
+                "_is_databricks_managed_embeddings",
+            )
+
+    return dependency_dict
+
+
+def _extract_databricks_dependencies_from_llm(llm, dependency_dict) -> Dict[str, str]:
+    import langchain
+
+    if isinstance(llm, langchain.llms.Databricks):
+        _assign_value_or_append_to_list(
+            dependency_dict, _DATABRICKS_LLM_ENDPOINT_NAME_KEY, llm.endpoint_name
+        )
+    return dependency_dict
+
+
+def _extract_databricks_dependencies_from_chat_model(chat_model, dependency_dict) -> Dict[str, str]:
+    import langchain
+
+    if isinstance(chat_model, langchain.chat_models.ChatDatabricks):
+        _assign_value_or_append_to_list(
+            dependency_dict, _DATABRICKS_CHAT_ENDPOINT_NAME_KEY, chat_model.endpoint
+        )
+    return dependency_dict
+
+
+def _extract_dependency_dict_from_lc_model(lc_model, dependency_dict) -> Dict[str, str]:
+    """
+    This function contains the logic to examine a non-Runnable component of a langchain model.
+    The logic here does not cover all legacy chains. If you need to support a custom chain,
+    you need to monkey patch this function.
+    """
+    if lc_model is None:
+        return dependency_dict
+
+    # leaf node
+    dependency_dict = _extract_databricks_dependencies_from_chat_model(lc_model, dependency_dict)
+    dependency_dict = _extract_databricks_dependencies_from_retriever(lc_model, dependency_dict)
+    dependency_dict = _extract_databricks_dependencies_from_llm(lc_model, dependency_dict)
+
+    # recursively inspect legacy chain
```

Comment:
Could we use a map then go over the map for this?
```
for key, att in map:
    if hasattr(lc_model, key):
        _extract_dependency_dict_from_lc_model(getattr(lc_model, att), dependency_dict)
```

Reply:
An irrelevant comment but since we're using dictionary, we don't really need to return the value as it's in-place updated